diff --git a/component/process/process_freebsd_amd64.go b/component/process/process_freebsd.go
similarity index 90%
rename from component/process/process_freebsd_amd64.go
rename to component/process/process_freebsd.go
index a6d00c8e..21094aac 100644
--- a/component/process/process_freebsd_amd64.go
+++ b/component/process/process_freebsd.go
@@ -177,23 +177,22 @@ func (s *searcher) searchSocketPid(socket uint64) (uint32, error) {
 	return 0, ErrNotFound
 }
 
-func newSearcher(major int) *searcher {
+func newSearcher(major int, machine string) *searcher {
 	var s *searcher
+	var vflag int
+	switch machine {
+	case "i386":
+		vflag = 388
+	case "amd64":
+		fallthrough
+	case "arm64":
+		fallthrough
+	case "riscv":
+		vflag = 392
+	default:
+		return nil
+	}
 	switch major {
-	case 11:
-		s = &searcher{
-			headSize:     32,
-			tcpItemSize:  1304,
-			udpItemSize:  632,
-			port:         198,
-			ip:           228,
-			vflag:        116,
-			socket:       88,
-			fileItemSize: 80,
-			data:         56,
-			pid:          8,
-			udpInpOffset: 8,
-		}
 	case 12:
 		fallthrough
 	case 13:
@@ -207,7 +206,7 @@ func newSearcher(major int) *searcher {
 			udpItemSize:  400,
 			port:         254,
 			ip:           284,
-			vflag:        392,
+			vflag:        vflag,
 			socket:       16,
 			fileItemSize: 128,
 			data:         56,
@@ -222,6 +221,10 @@ func initSearcher() error {
 	if err != nil {
 		return err
 	}
+	machine, err := syscall.Sysctl("hw.machine")
+	if err != nil {
+		return err
+	}
 
 	dot := strings.Index(osRelease, ".")
 	if dot != -1 {
@@ -231,9 +234,9 @@ func initSearcher() error {
 	if err != nil {
 		return err
 	}
-	defaultSearcher = newSearcher(major)
+	defaultSearcher = newSearcher(major, machine)
 	if defaultSearcher == nil {
-		return fmt.Errorf("unsupported freebsd version %d", major)
+		return fmt.Errorf("unsupported freebsd version %d (%s)", major, machine)
 	}
 	return nil
 }
diff --git a/component/process/process_other.go b/component/process/process_other.go
index eea6e5fd..83b6f085 100644
--- a/component/process/process_other.go
+++ b/component/process/process_other.go
@@ -1,4 +1,4 @@
-//go:build !darwin && !linux && !windows && (!freebsd || !amd64)
+//go:build !darwin && !linux && !windows && !freebsd
 
 package process
 
diff --git a/listener/config/tun.go b/listener/config/tun.go
index 8c125b66..8f33bbf5 100644
--- a/listener/config/tun.go
+++ b/listener/config/tun.go
@@ -24,6 +24,7 @@ type Tun struct {
 	Inet6Address                          []netip.Prefix `yaml:"inet6-address" json:"inet6-address,omitempty"`
 	IPRoute2TableIndex                    int            `yaml:"iproute2-table-index" json:"iproute2-table-index,omitempty"`
 	IPRoute2RuleIndex                     int            `yaml:"iproute2-rule-index" json:"iproute2-rule-index,omitempty"`
+	FIBIndex                              int            `yaml:"fib-index" json:"fib-index,omitempty"`
 	AutoRedirect                          bool           `yaml:"auto-redirect" json:"auto-redirect,omitempty"`
 	AutoRedirectInputMark                 uint32         `yaml:"auto-redirect-input-mark" json:"auto-redirect-input-mark,omitempty"`
 	AutoRedirectOutputMark                uint32         `yaml:"auto-redirect-output-mark" json:"auto-redirect-output-mark,omitempty"`
@@ -128,6 +129,9 @@ func (t *Tun) Equal(other Tun) bool {
 	if t.IPRoute2RuleIndex != other.IPRoute2RuleIndex {
 		return false
 	}
+	if t.FIBIndex != other.FIBIndex {
+		return false
+	}
 	if t.AutoRedirect != other.AutoRedirect {
 		return false
 	}
diff --git a/listener/listener.go b/listener/listener.go
index 37bcbebf..cb74ea97 100644
--- a/listener/listener.go
+++ b/listener/listener.go
@@ -3,6 +3,7 @@ package listener
 import (
 	"fmt"
 	"net"
+	"runtime"
 	"strconv"
 	"strings"
 	"sync"
@@ -234,9 +235,11 @@ func ReCreateRedir(port int, tunnel C.Tunnel) {
 		return
 	}
 
-	redirUDPListener, err = tproxy.NewUDP(addr, tunnel)
-	if err != nil {
-		log.Warnln("Failed to start Redir UDP Listener: %s", err)
+	if runtime.GOOS != "freebsd" {
+		redirUDPListener, err = tproxy.NewUDP(addr, tunnel)
+		if err != nil {
+			log.Warnln("Failed to start Redir UDP Listener: %s", err)
+		}
 	}
 
 	log.Infoln("Redirect proxy listening at: %s", redirListener.Address())
diff --git a/listener/redir/tcp_freebsd.go b/listener/redir/tcp_freebsd.go
index 9eb199f0..0a3199a3 100644
--- a/listener/redir/tcp_freebsd.go
+++ b/listener/redir/tcp_freebsd.go
@@ -5,62 +5,87 @@ import (
 	"errors"
 	"net"
 	"net/netip"
+	"os"
 	"syscall"
 	"unsafe"
 
 	"github.com/metacubex/mihomo/transport/socks5"
-
-	"golang.org/x/sys/unix"
 )
 
 const (
-	SO_ORIGINAL_DST      = 80 // from linux/include/uapi/linux/netfilter_ipv4.h
-	IP6T_SO_ORIGINAL_DST = 80 // from linux/include/uapi/linux/netfilter_ipv6/ip6_tables.h
+	PF_IN       = 0x1        // contrib/tcpdump/pflog.h
+	PF_OUT      = 0x2        // contrib/tcpdump/pflog.h
+	DIOCNATLOOK = 0xc04c4417 // sys/net/pfvar.h
 )
 
 func parserPacket(conn net.Conn) (socks5.Addr, error) {
-	c, ok := conn.(*net.TCPConn)
+	_, ok := conn.(*net.TCPConn)
 	if !ok {
 		return nil, errors.New("only work with TCP connection")
 	}
 
-	rc, err := c.SyscallConn()
-	if err != nil {
-		return nil, err
-	}
-
-	var addr netip.AddrPort
-
-	rc.Control(func(fd uintptr) {
-		if ip4 := c.LocalAddr().(*net.TCPAddr).IP.To4(); ip4 != nil {
-			addr, err = getorigdst(fd)
+	_, err := os.Stat("/dev/pf")
+	if errors.Is(err, os.ErrNotExist) {
+		// ipfw
+		la := conn.LocalAddr().(*net.TCPAddr)
+		var ip net.IP
+		if la.IP.To4() != nil {
+			ip = make(net.IP, net.IPv4len)
+			copy(ip, la.IP[net.IPv6len-net.IPv4len:])
 		} else {
-			addr, err = getorigdst6(fd)
+			ip = make(net.IP, net.IPv6len)
+			copy(ip, la.IP[:])
 		}
-	})
+		addr, _ := netip.AddrFromSlice(ip)
+		return socks5.AddrFromStdAddrPort(netip.AddrPortFrom(addr, uint16(la.Port))), nil
+	}
 
-	return socks5.AddrFromStdAddrPort(addr), err
-}
+	// pf
+	fd, err := syscall.Open("/dev/pf", 0, syscall.O_RDWR)
+	if err != nil {
+		return nil, err
+	}
+	defer syscall.Close(fd)
 
-// Call getorigdst() from linux/net/ipv4/netfilter/nf_conntrack_l3proto_ipv4.c
-func getorigdst(fd uintptr) (netip.AddrPort, error) {
-	addr := unix.RawSockaddrInet4{}
-	size := uint32(unsafe.Sizeof(addr))
-	_, _, err := syscall.Syscall6(syscall.SYS_GETSOCKOPT, fd, syscall.IPPROTO_IP, SO_ORIGINAL_DST, uintptr(unsafe.Pointer(&addr)), uintptr(unsafe.Pointer(&size)), 0)
-	if err != 0 {
-		return netip.AddrPort{}, err
+	nl := struct {
+		saddr, daddr, rsaddr, rdaddr [16]byte
+		sport, dport, rsport, rdport [2]byte
+		af, proto, direction         uint8
+	}{
+		af:        syscall.AF_INET,
+		proto:     syscall.IPPROTO_TCP,
+		direction: PF_OUT,
+	}
+	la := conn.LocalAddr().(*net.TCPAddr)
+	ra := conn.RemoteAddr().(*net.TCPAddr)
+	raIP, laIP := ra.IP, la.IP
+	raPort, laPort := ra.Port, la.Port
+	switch {
+	case raIP.To4() != nil:
+		copy(nl.saddr[:net.IPv4len], raIP.To4())
+		copy(nl.daddr[:net.IPv4len], laIP.To4())
+		nl.af = syscall.AF_INET
+	default:
+		copy(nl.saddr[:], raIP.To16())
+		copy(nl.daddr[:], laIP.To16())
+		nl.af = syscall.AF_INET6
+	}
+	binary.BigEndian.PutUint16(nl.sport[:], uint16(raPort))
+	binary.BigEndian.PutUint16(nl.dport[:], uint16(laPort))
+	if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, uintptr(fd), DIOCNATLOOK, uintptr(unsafe.Pointer(&nl))); errno != 0 {
+		return nil, errno
 	}
-	port := binary.BigEndian.Uint16((*(*[2]byte)(unsafe.Pointer(&addr.Port)))[:])
-	return netip.AddrPortFrom(netip.AddrFrom4(addr.Addr), port), nil
-}
 
-func getorigdst6(fd uintptr) (netip.AddrPort, error) {
-	addr := unix.RawSockaddrInet6{}
-	size := uint32(unsafe.Sizeof(addr))
-	_, _, err := syscall.Syscall6(syscall.SYS_GETSOCKOPT, fd, syscall.IPPROTO_IPV6, IP6T_SO_ORIGINAL_DST, uintptr(unsafe.Pointer(&addr)), uintptr(unsafe.Pointer(&size)), 0)
-	if err != 0 {
-		return netip.AddrPort{}, err
+	var ip net.IP
+	switch nl.af {
+	case syscall.AF_INET:
+		ip = make(net.IP, net.IPv4len)
+		copy(ip, nl.rdaddr[:net.IPv4len])
+	case syscall.AF_INET6:
+		ip = make(net.IP, net.IPv6len)
+		copy(ip, nl.rdaddr[:])
 	}
-	port := binary.BigEndian.Uint16((*(*[2]byte)(unsafe.Pointer(&addr.Port)))[:])
-	return netip.AddrPortFrom(netip.AddrFrom16(addr.Addr), port), nil
+	port := binary.BigEndian.Uint16(nl.rdport[:])
+	addr, _ := netip.AddrFromSlice(ip)
+	return socks5.AddrFromStdAddrPort(netip.AddrPortFrom(addr, port)), nil
 }
diff --git a/listener/sing_tun/server.go b/listener/sing_tun/server.go
index 56369945..cdd6db83 100644
--- a/listener/sing_tun/server.go
+++ b/listener/sing_tun/server.go
@@ -207,6 +207,10 @@ func New(options LC.Tun, tunnel C.Tunnel, additions ...inbound.Addition) (l *Lis
 	if ruleIndex == 0 {
 		ruleIndex = tun.DefaultIPRoute2RuleIndex
 	}
+	fibIndex := options.FIBIndex
+	if fibIndex == 0 {
+		fibIndex = tun.DefaultFIBIndex
+	}
 	autoRedirectFallbackRuleIndex := options.AutoRedirectIPRoute2FallbackRuleIndex
 	if autoRedirectFallbackRuleIndex == 0 {
 		autoRedirectFallbackRuleIndex = tun.DefaultIPRoute2AutoRedirectFallbackRuleIndex
@@ -372,6 +376,7 @@ func New(options LC.Tun, tunnel C.Tunnel, additions ...inbound.Addition) (l *Lis
 		IPRoute2TableIndex:                    tableIndex,
 		IPRoute2RuleIndex:                     ruleIndex,
 		IPRoute2AutoRedirectFallbackRuleIndex: autoRedirectFallbackRuleIndex,
+		FIBIndex:                              fibIndex,
 		AutoRedirectInputMark:                 inputMark,
 		AutoRedirectOutputMark:                outputMark,
 		Inet4LoopbackAddress:                  common.Filter(options.LoopbackAddress, netip.Addr.Is4),
